docker/main/ngen/py_funcs.py

```diff
@@ -0,0 +1,310 @@
+    """
+    Make a tar archive from the contents of a directory, and place this in a specified destination.
+
+    Parameters
+    ----------
+    source
+        Directory whose contents should be archived.
+    dest
+        Destination directory in which to place the archive file.
+    archive_name
+        Basename for the created archive file.
+    do_dry_run
+        Whether to only perform a dry run to check paths, with no archiving/moving/copying.
+    do_compress
+        Whether to compress the created archive with gzip compression.
+
+    Raises
+    -------
+    ValueError
+        If the source or dest directory either does not exist or is not a directory; also, if the archive file already
+        exists.
+    """
+    if not source.exists():
+        raise ValueError(f"{get_date_str()} tar_and_copy source directory {source!s} does not exist!")
+    elif not source.is_dir():
+        raise ValueError(f"{get_date_str()} Non-directory file at path for tar_and_copy source directory {source!s}!")
+
+    if not dest.exists():
+        raise ValueError(f"{get_date_str()} tar_and_copy destination directory {dest!s} does not exist!")
+    elif not dest.is_dir():
+        raise ValueError(f"{get_date_str()} Non-directory file at path for tar_and_copy dest directory {dest!s}!")
+
+    # We may change things in the future to write somewhere else, so do things in a little more of a round-about way
+    # This is where the archive initially gets created
+    archive_create_path = dest.joinpath(archive_name)
+    # This is where it must eventually go
+    final_archive_path = dest.joinpath(archive_name)
+
+    if archive_create_path.exists():
+        raise ValueError(f"{get_date_str()} File exists at tar_and_copy archive path {archive_create_path!s}!")
+    if final_archive_path != archive_create_path and final_archive_path.exists():
+        raise ValueError(f"{get_date_str()} Archive for tar_and_copy already exists in dest at {final_archive_path!s}!")
+
+    if do_dry_run:
+        return
+
+    tar_mode_args = "w:gz" if do_compress else "w"
+    with tarfile.open(archive_create_path, tar_mode_args) as tar:
+        for p in source.glob("*"):
+            tar.add(p, arcname=p.name)
+
+    if archive_create_path != final_archive_path:
+        shutil.move(archive_create_path, final_archive_path)
+
+
+def main():
+    args = _parse_args()
+
+    _apply_logging(args.log_level)
+
+    if args.command == 'tar_and_copy':
+        tar_and_copy(**(vars(args)))
+    elif args.command == 'gather_output':
+        mpi_host_to_nproc_map = process_mpi_hosts_string(args.mpi_hosts_str)
+        gather_output(mpi_host_names=[h for h in mpi_host_to_nproc_map], output_write_dir=args.output_write_dir)
+    elif args.command == 'move_job_output':
+        move_job_output(**(vars(args)))
+    else:
+        raise ValueError("Unsupported command {}".format(args.command))
```

Comment:
Strictly speaking, yes, but I like this for future proofing.  I'm one to occasionally forget the "connecting wiring" between things, like adding the lines in `main` between the arg parsing for a new function and the function implementation itself.

Would it be more appropriate if I changed this to `NotImplementedError`?

Reply:
Yeah - `NotImplementedError` is more appropriate here. Adding a whole sub parser and not handling it in `main` is a big no-no and should be caught in PR, but this is perhaps the only way to catch a mess up of that magnitude.

That scenario is also why subparsers should be avoided, though.

Reply:
Eh, actually, I went and read the documentation for _NotImplementedError_ after suggesting it.  I'm not sure it's more appropriate here:

> In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method, or while the class is being developed to indicate that the real implementation still needs to be added.

Aside from outright forgetting, raising some kind of exception here also protects against typos in the compared-to literal strings in the above expressions; otherwise, things would just silently not run (but also not actually fail).  I've switched this to a _RuntimeError_ and modified the message a bit.